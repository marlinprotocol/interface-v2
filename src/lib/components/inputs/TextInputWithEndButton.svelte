<script lang="ts">
	import { inputClasses } from '$lib/atoms/componentClasses';
	import InputCardWithEndButton from '$lib/components/inputs/InputCardWithEndButton.svelte';
	import { cn } from '$lib/utils/helpers/commonHelper';

	export let styleClass = '';
	export let title = '';
	export let placeholder = '';
	export let disabled = false;
	export let input: string;

	const titleStyleInit =
		'absolute cursor-text px-1 font-poppins text-sm text-[#030115] transition-all';
	const onBlurStyle = 'top-[50%] left-10% font-medium bg-transparent translate-y-[-50%]';
	const onFocusStyle = 'left-10% top-[-10px] hidden font-normal bg-white';
	let titleStyle = onFocusStyle;
	export let label: string = '';
	export let showTitle: boolean = false;

	const onFocus = () => {
		titleStyle = onFocusStyle;
	};
	const onBlur = () => {
		if (!input.length) {
			titleStyle = onBlurStyle;
		} else {
			onFocus();
		}
	};
	if (!disabled) {
		titleStyle = onBlurStyle;
	}
	$: if (!!input.length) {
		titleStyle = onFocusStyle;
	}
</script>

<InputCardWithEndButton
	{showTitle}
	{label}
	variant="v2Input"
	{title}
	styleClass={cn(styleClass, 'relative')}
>
	{#if !!title.length}
		<button class={cn(titleStyle, titleStyleInit)}>
			{title}
		</button>
	{/if}
	<input
		id="address-display"
		bind:value={input}
		on:focus={onFocus}
		on:blur={onBlur}
		class={cn(inputClasses.inputText, 'hideInputNumberAppearance')}
		{placeholder}
		{disabled}
	/>
	<svelte:fragment slot="endInfoBox">
		<slot name="endInfoBox" />
	</svelte:fragment>
	<svelte:fragment slot="titleEndButton">
		<slot name="titleEndButton" />
	</svelte:fragment>
	<svelte:fragment slot="endButton">
		<slot name="endButton" />
	</svelte:fragment>
</InputCardWithEndButton>
