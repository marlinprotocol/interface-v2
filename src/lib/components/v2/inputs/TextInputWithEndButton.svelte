<script lang="ts">
	import { inputClasses } from '$lib/atoms/v2/componentClasses';
	import InputCardWithEndButton from '$lib/components/v2/inputs/InputCardWithEndButton.svelte';
	import { cn } from '$lib/utils/helpers/commonHelper';

	export let styleClass = '';
	export let title = '';
	export let placeholder = '';
	export let disabled = false;
	export let input: string;
	export let id: string;

	const titleStyleInit =
		'absolute cursor-text px-1 font-poppins text-sm text-[#030115] transition-all';
	const onBlurStyle = 'top-[50%] left-10% font-medium bg-transparent translate-y-[-50%]';
	const onFocusStyle = 'left-10% top-[-10px] hidden font-light bg-white';
	let isFocused = false;
	let titleStyle = onFocusStyle;
	export let label: string = '';
	export let showTitle: boolean = false;

	function focusElement() {
		const eleToFocus: HTMLElement | null = document.getElementById(`text-input-title-${id}`);
		if (eleToFocus) {
			eleToFocus.focus();
		}
	}
	const onFocus = () => {
		isFocused = true;
		titleStyle = onFocusStyle;
	};
	const onBlur = () => {
		if (!input.length) {
			isFocused = false;
			titleStyle = onBlurStyle;
		} else {
			onFocus();
		}
	};
	if (!disabled) {
		titleStyle = onBlurStyle;
	}
	$: if (!!input.length) {
		titleStyle = onFocusStyle;
	}
</script>

<InputCardWithEndButton
	{showTitle}
	{label}
	variant="v2Input"
	{title}
	styleClass={cn(styleClass, 'relative')}
>
	{#if !!title.length}
		<button
			class={cn(titleStyle, titleStyleInit)}
			id="text-input-title-{id}"
			on:click={focusElement}
		>
			{title}
		</button>
	{/if}
	<input
		id="address-display"
		bind:value={input}
		on:focus={onFocus}
		on:blur={onBlur}
		class={cn(inputClasses.inputText, 'hideInputNumberAppearance')}
		placeholder={isFocused ? placeholder : ''}
		{disabled}
	/>
	<svelte:fragment slot="endInfoBox">
		<slot name="endInfoBox" />
	</svelte:fragment>
	<svelte:fragment slot="titleEndButton">
		<slot name="titleEndButton" />
	</svelte:fragment>
	<svelte:fragment slot="endButton">
		<slot name="endButton" />
	</svelte:fragment>
</InputCardWithEndButton>
