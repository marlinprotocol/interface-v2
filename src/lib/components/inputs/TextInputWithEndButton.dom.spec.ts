import { render } from '@testing-library/svelte';
import { describe, it, expect } from 'vitest';
import TextInputWithEndButton from './TextInputWithEndButton.svelte';
import html from '@playpilot/svelte-htm';

describe('TextInputWithEndButton', () => {
	const tooltipText = 'Test Tooltip';
	const title = 'Test Title';
	const placeholder = 'Enter something...';
	const styleClass = 'test-style';
	const inputValue = 'Test input value';

	it("render's the component properly", () => {
		const { container } = render(TextInputWithEndButton, {
			styleClass,
			title,
			placeholder,
			input: inputValue,
			disabled: false
		});
		expect(container).toMatchSnapshot();
	});

	it('passes props correctly to InputCardWithEndButton and renders input elements', () => {
		const { getByPlaceholderText, getByText } = render(TextInputWithEndButton, {
			styleClass,
			title,
			placeholder,
			input: inputValue,
			disabled: false
		});
		// Assert input properties are bound correctly
		const inputElement = getByPlaceholderText(placeholder) as HTMLInputElement;
		expect(inputElement.value).toBe(inputValue);
		expect(inputElement.disabled).toBeFalsy();
	});

	it("render's slots correctly", () => {
		const { getByText } = render(html`
        <${TextInputWithEndButton} 
        styleClass=${styleClass}
        tooltipText=${tooltipText} 
        title=${title} 
        placeholder=${placeholder}
        input=${inputValue}
        disabled=${false}
        >
         <div slot='titleEndButton'>titleEndButton</div>
         <div slot='endButton'>endButton</div>
        </${TextInputWithEndButton}>
      `);
		// Assert that slot content is rendered
		expect(getByText('titleEndButton')).toBeTruthy();
		expect(getByText('endButton')).toBeTruthy();
	});

	it('disables the input when disabled prop is true', () => {
		const placeholder = 'Enter something...';
		const { getByPlaceholderText } = render(TextInputWithEndButton, {
			styleClass,
			title,
			placeholder,
			input: inputValue,
			disabled: true
		});

		// Find the input by its placeholder and assert it's disabled
		const inputElement = getByPlaceholderText(placeholder) as HTMLInputElement;
		expect(inputElement.disabled).toBeTruthy();
	});

	// ... more tests for other props and behaviors ...
});
