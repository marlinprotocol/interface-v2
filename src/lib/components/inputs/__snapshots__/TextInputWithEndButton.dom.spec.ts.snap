// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextInputWithEndButton > render's the component properly 1`] = `
<body>
  <div>
    <div
      class="w-full border border-grey-100 bg-base-100 py-2.5 px-4 rounded-full test-style relative"
      data-testid="input-card"
    >
       
      <div
        class="flex items-center justify-between"
      >
         
      </div>
       
      <div
        class="flex items-center justify-between"
      >
        <input
          class="input input-ghost h-[30px] w-full p-0 font-normal text-grey-800 disabled:opacity-70 disabled:text-grey-800 disabled:placeholder:text-grey-800/[.2] focus-within:text-grey-800 placeholder:text-grey-800/[.2] focus:outline-none focus-within:border-b-2 focus:bg-transparent hideInputNumberAppearance"
          id="address-display"
          placeholder="Enter something..."
        />
         
         
      </div>
    </div>
    <!--&lt;InputCard&gt;-->
    <!--&lt;InputCardWithEndButton&gt;-->
    <!--&lt;TextInputWithEndButton&gt;-->
  </div>
</body>
`;
